## Gather

In [1]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name    surname                address  \
63           64         male   Yukitaka   Takenaka     1975 Holden Street   
24           25         male      Jakob   Jakobsen      648 Old Dear Lane   
128         129         male   Muhammad     Hughes   1965 Crestview Manor   
99          100         male      Bernd  Schneider        1262 Deans Lane   
48           49       female     Annika      Vaara  4853 University Drive   

             city       state  zip_code        country  \
63      San Diego  California   92103.0  United States   
24    Port Jervis    New York   12771.0  United States   
128  Indianapolis          IN   46214.0  United States   
99       Westbury    New York   11590.0  United States   
48        Lombard          IL   60148.0  United States   

                                        contact   birthdate  weight  height  \
63      619-299-1495YukitakaTakenaka@einrot.com  11/24/1944   193.2      72   
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
128      MuhammadHughes@rhyta.com1 317 292 2394   9/21/1938   202.2      72   
99    BerndSchneider@jourrapide.com914-830-3940   3/24/1993   212.5      71   
48            AnnikaVaara@rhyta.com312-719-7238   2/20/1996   109.1      67   

      bmi  
63   26.2  
24   24.4  
128  27.4  
99   29.6  
48   17.1

In [15]:
patients.surname.value_counts()

Doe          6
Taylor       3
Jakobsen     3
Tạ           2
Tucker       2
            ..
Terrazas     1
Ekwueme      1
Šlezinger    1
Azevedo      1
Brodahl      1
Name: surname, Length: 466, dtype: int64

In [16]:
patients.address.value_counts()

123 Main Street            6
2778 North Avenue          2
648 Old Dear Lane          2
2476 Fulton Street         2
1510 Allison Avenue        1
                          ..
1742 Tea Berry Lane        1
4278 Hart Country Lane     1
4155 Raccoon Run           1
4110 Capitol Avenue        1
3261 Desert Broom Court    1
Name: address, Length: 483, dtype: int64

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame..*

##### Code

In [24]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean , treatments_cut] , ignore_index=True)

##### Test

In [25]:
# Your testing code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Recalculate `hba1c_change` = `hba1c_start` - `hba1c_end` .*

##### Code

In [26]:
# Your cleaning code here
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [27]:
# Your testing code here
treatments_clean['hba1c_change'].values

array([0.43, 0.47, 0.43, 0.35, 0.32, 0.38, 0.38, 0.34, 0.38, 0.39, 0.51,
       0.38, 0.4 , 0.45, 0.44, 0.44, 0.4 , 0.48, 0.42, 0.41, 0.42, 0.32,
       0.43, 0.42, 0.39, 0.42, 0.44, 0.47, 0.39, 0.33, 0.32, 0.47, 0.49,
       0.36, 0.41, 0.4 , 0.44, 0.45, 0.35, 0.51, 0.46, 0.42, 0.38, 0.39,
       0.3 , 0.35, 0.35, 0.45, 0.4 , 0.48, 0.43, 0.54, 0.45, 0.45, 0.41,
       0.37, 0.47, 0.45, 0.4 , 0.42, 0.46, 0.31, 0.38, 0.33, 0.38, 0.32,
       0.31, 0.46, 0.32, 0.39, 0.24, 0.47, 0.39, 0.48, 0.4 , 0.39, 0.42,
       0.42, 0.34, 0.5 , 0.54, 0.32, 0.49, 0.38, 0.34, 0.38, 0.51, 0.39,
       0.41, 0.42, 0.34, 0.43, 0.42, 0.29, 0.4 , 0.31, 0.45, 0.37, 0.44,
       0.38, 0.39, 0.38, 0.41, 0.4 , 0.37, 0.34, 0.47, 0.37, 0.42, 0.44,
       0.46, 0.28, 0.42, 0.39, 0.37, 0.41, 0.42, 0.41, 0.32, 0.36, 0.36,
       0.46, 0.39, 0.37, 0.51, 0.37, 0.42, 0.4 , 0.44, 0.36, 0.38, 0.34,
       0.44, 0.43, 0.33, 0.38, 0.51, 0.44, 0.49, 0.44, 0.37, 0.3 , 0.35,
       0.35, 0.27, 0.33, 0.41, 0.46, 0.32, 0.46, 0.

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*extract phone number and email and store them in separate variables(columns).*

##### Code

In [28]:
# Your cleaning code here
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)


In [29]:
##patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [30]:
# Your testing code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number                      email  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*put auralin & novodra in `treatment` column
make start- end dose in `dose` column
then split `dose` column into `start` and `end` dose .*

##### Code

In [31]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [32]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean , 
                          id_vars = ['given_name' , 'surname' , 'hba1c_start' , 'hba1c_end' , 'hba1c_change'],
                          var_name='treatment' , value_name='dose')

In [33]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika    jindrová         7.63       7.20          0.43   auralin   
1       elliot  richardson         7.56       7.09          0.47   auralin   
2     yukitaka    takenaka         7.68       7.25          0.43   auralin   
3         skye  gormanston         7.97       7.62          0.35   auralin   
4       alissa      montez         7.78       7.46          0.32   auralin   
..         ...         ...          ...        ...           ...       ...   
695     rovzan     kishiev         7.75       7.41          0.34   novodra   
696      jakob    jakobsen         7.96       7.51          0.45   novodra   
697      bernd   schneider         7.74       7.44          0.30   novodra   
698      berta  napolitani         7.68       7.21          0.47   novodra   
699     armina       sauvé         7.86       7.40          0.46   novodra   

          dose  
0    41u - 48u  
1            -  
2            -  
3    33u - 36u  
4            -  
..         ...  
695          -  
696  28u - 26u  
697          -  
698  42u - 44u  
699          -  

[700 rows x 7 columns]

In [136]:
treatments_clean = treatments_clean[treatments_clean.dose != "-"]

In [137]:
treatments_clean['dose_start'] , treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ' , 1).str

In [138]:
treatments_clean = treatments_clean.drop('dose' , axis=1)

##### Test

In [139]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
*merge `adverse_reaction` column to treatments table on the `surname` and `given_name`.*

##### Code

In [140]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean , adverse_reactions_clean, 
                           on = ['given_name','surname'],
                           how = 'left')

##### Test

In [141]:
# Your testing code here
treatments_clean

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        veronika       jindrová         7.63       7.20          0.43   
1            skye     gormanston         7.97       7.62          0.35   
2          sophia         haugen         7.65       7.27          0.38   
3           eddie         archer         7.89       7.55          0.34   
4            asia        woźniak         7.76       7.37          0.39   
5          joseph            day         7.70       7.19          0.51   
6         roxanne      andreyeva         9.54       9.14          0.40   
7          simone   baumgaertner         7.74       7.30          0.44   
8            enco         žibrik         7.78       7.34          0.44   
9         camilla       zaitseva         7.53       7.13          0.40   
10          tekla        walczak         7.61       7.29          0.32   
11    brancaleone          russo         8.61       8.18          0.43   
12           isac           berg         9.68       9.29          0.39   
13        clinton         miller         7.79       7.40          0.39   
14         eugene        mironov         7.81       7.48          0.33   
15    szilveszter          totth         7.70       7.38          0.32   
16      alexander      mathiesen         7.96       7.55          0.41   
17            chỉ            lâm         7.68       7.24          0.44   
18      władysław      wieczorek         7.92       7.47          0.45   
19       kristján        ingason         7.92       7.57          0.35   
20         marija       grubišić         7.53       7.15          0.38   
21          sauli     koivuniemi         7.67       7.37          0.30   
22        mariana          souza         7.86       7.51          0.35   
23     kristoffer      martinsen         9.18       8.64          0.54   
24             mỹ          quynh         7.61       7.16          0.45   
25           oles        zhdanov         7.52       7.11          0.41   
26        triana.       terrazas         7.71       7.34          0.37   
27         gabryŝ    tomaszewski         7.87       7.47          0.40   
28      leixandre         alanis         7.74       7.32          0.42   
29   onyekachukwu         obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
320          jane        citizen         7.98       7.60          0.38   
321        angela     lavrentyev         7.61       7.14          0.47   
322        edelma    villalpando         7.99       7.56          0.43   
323        annika          vaara         7.73       7.34          0.39   
324         chiho           higa         7.71       7.30          0.41   
325     beatrycze        woźniak         7.54       7.17          0.37   
326      miłosław     wiśniewski         7.51       7.08          0.43   
327       firenze          fodor         7.89       7.55          0.34   
328           zoe        wellish         7.71       7.30          0.41   
329           una  traustadóttir         8.00       7.50          0.50   
330         luboš          pecha         7.79       7.45          0.34   
331         meaza         brhane         7.70       7.36          0.34   
332         adlan       shishani         7.84       7.37          0.47   
333         sofia      hermansen         8.90       8.57          0.33   
334         guðni      heimisson         7.64       7.24          0.40   
335       eufemio        rosario         7.54       7.26          0.28   
336      dalmacia         madrid         7.67       7.21          0.46   
337         daimy          tromp         9.41       8.94          0.47   
338        jeremy        montagu         7.68       7.36          0.32   
339       nebechi      ekechukwu         7.78       7.39          0.39   
340       satsita      batukayev         7.63       7.25          0.38   
341       timothy         cotton         7.92       7.52          0.40   
342       b

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*isolate `patient_id` and names to convert to lower case and join with the treatments table then drop the names columns from the treatments df .*

##### Code

In [142]:
# Your cleaning code here
id_names = patients_clean[['patient_id' , 'given_name',  'surname']]
id_names['given_name'] = id_names['given_name'].str.lower()
id_names['surname'] = id_names['surname'].str.lower()
treatments_clean = pd.merge(treatments_clean , id_names , on =['given_name','surname'])
treatments_clean = treatments_clean.drop(['given_name','surname'] , axis=1)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


##### Test

In [143]:
# Your testing code here
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin        41u      48u   
1         7.97       7.62          0.35   auralin        33u      36u   
2         7.65       7.27          0.38   auralin        37u      42u   
3         7.89       7.55          0.34   auralin        31u      38u   
4         7.76       7.37          0.39   auralin        30u      36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*convert zip code into string and cut the extra zero out .*

##### Code

In [144]:
# Your cleaning code here
patients_clean['zip_code'] = patients_clean['zip_code'].astype(str).str[:-2].str.pad(5 , fillchar = '0')
patients_clean['zip_code'] = patients_clean['zip_code'].replace('0000n', np.nan)

##### Test

In [145]:
# Your testing code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  \
0  Rancho California  California    92390  United States   
1          Armstrong    Illinois    61812  United States   
2               York    Nebraska    68467  United States   
3         Woodbridge          NJ    07095  United States   
4             Dothan          AL    36303  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number                      email  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*replace height in rows with height = 27 to height = 72.*

##### Code

In [146]:
# Your cleaning code here
patients_clean['height'] = patients_clean['height'].replace(27 , 72)

##### Test

In [147]:
# Your testing code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  \
0  Rancho California  California    92390  United States   
1          Armstrong    Illinois    61812  United States   
2               York    Nebraska    68467  United States   
3         Woodbridge          NJ    07095  United States   
4             Dothan          AL    36303  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      72   

    bmi       phone_number                      email  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
*define function to replace avery state full name into the abbreviation .*

##### Code

In [148]:
# Your cleaning code here
state_abbrv = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE' }

def abbreviate_state(patient):
    if patient['state'] in state_abbrv.keys():
        abbrv = state_abbrv[patient['state']]
        return abbrv
    else:
        return patient['state']
    
patients_clean['state']= patients_clean.apply(abbreviate_state , axis=1)

##### Test

In [149]:
# Your testing code here
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
MI    13
OK    13
LA    13
NJ    12
VA    11
MS    10
WI    10
MN     9
AL     9
IN     9
TN     9
KY     8
WA     8
NC     8
MO     7
KS     6
ID     6
NE     6
NV     6
IA     5
CT     5
SC     5
ND     4
AR     4
AZ     4
RI     4
ME     4
CO     4
WV     3
MD     3
SD     3
OR     3
DE     3
MT     2
VT     2
DC     2
NM     1
WY     1
NH     1
AK     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
*fix the name spelling david.*

##### Code

In [150]:
# Your cleaning code here
patients_clean['given_name'] = patients_clean['given_name'].replace('Dsvid' , 'David')

##### Test

In [153]:
# Your testing code here
patients_clean[patients_clean['given_name'] == 'Dsvid']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, contact, birthdate, weight, height, bmi, phone_number, email]
Index: []

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*convert `assigned_sex` and `state` to category types 
convert `birthdate` to datetime type and convert start and end dose to integers and cut the "u" letter out .*

##### Code

In [155]:
# Your cleaning code here
patients_clean[['assigned_sex' , 'state']] = patients_clean[['assigned_sex' , 'state']].astype('category')

patients_clean['birthdate']  = pd.to_datetime(patients_clean['birthdate'])

treatments_clean['dose_start'] = treatments_clean['dose_start'].str.strip('u').astype(int)
treatments_clean['dose_end'] = treatments_clean['dose_end'].str.strip('u').astype(int)

##### Test

In [156]:
# Your testing code here
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 16 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(9)
memory usage: 57.8+ KB


In [157]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


In [158]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment  dose_start  dose_end  \
0         7.63       7.20          0.43   auralin          41        48   
1         7.97       7.62          0.35   auralin          33        36   
2         7.65       7.27          0.38   auralin          37        42   
3         7.89       7.55          0.34   auralin          31        38   
4         7.76       7.37          0.39   auralin          30        36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

#### Multiple phone number formats

##### Define
*strip ' ' '-' '+'',' and pad the number with 1.*

##### Code

In [159]:
# Your cleaning code here
patients_clean['phone_number'] = patients_clean['phone_number'].str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [160]:
# Your testing code here
patients_clean['phone_number'].head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
*remove all rows with `surname` = Doe.*

##### Code

In [161]:
# Your cleaning code here
patients_clean = patients_clean[patients_clean['surname'] != 'Doe']

##### Test

In [162]:
# Your testing code here
patients_clean['surname'].value_counts()

Taylor           3
Jakobsen         3
Tạ               2
Tucker           2
Collins          2
Woźniak          2
Grímsdóttir      2
Lund             2
Lâm              2
Souza            2
Silva            2
Bùi              2
Dratchev         2
Correia          2
Parker           2
Batukayev        2
Aranda           2
Nilsen           2
Hueber           2
Johnson          2
Kadyrov          2
Liễu             2
Ogochukwu        2
Lương            2
Cabrera          2
Cindrić          2
Berg             2
Kowalczyk        2
Gersten          2
Schiavone        2
                ..
Borgen           1
Madrid           1
Pecinová         1
Guzman           1
Beauvais         1
Abraham          1
Vũ               1
Reilly           1
Prince           1
Sung             1
Nakayama         1
Filípek          1
Sakai            1
Blix             1
Murakami         1
Nowakowski       1
Tobeolisa        1
Nielsen          1
Michalčíková     1
Chinedum         1
Mattila          1
Stanić      

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [163]:
# Your cleaning code here
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [164]:
# Your testing code here
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  \
24   Port Jervis    NY    12771  United States   
432      Houston    TX    77020  United States   

                                        contact  birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707 1985-08-01   155.8      67   
432  KarenJakobsen@jourrapide.com1 979 203 0438 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

#### kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [165]:
# Your cleaning code here
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

##### Test

In [166]:
# Your testing code here
patients_clean.weight.sort_values()

459    102.100000
335    102.700000
74     103.200000
317    106.000000
171    106.500000
51     107.100000
210    107.585456
270    108.100000
198    108.500000
48     109.100000
478    109.600000
141    110.200000
38     111.800000
438    112.000000
14     112.000000
235    112.200000
307    112.400000
191    112.600000
408    113.100000
49     113.300000
326    114.000000
338    114.100000
253    117.000000
321    118.400000
168    118.800000
1      118.800000
350    119.000000
207    119.200000
265    120.000000
341    120.300000
          ...    
332    224.000000
12     224.200000
252    224.200000
222    224.800000
166    225.300000
111    225.900000
101    226.200000
150    226.600000
88     227.700000
352    227.700000
428    227.700000
13     228.400000
339    229.000000
182    230.300000
121    230.800000
257    231.700000
395    231.900000
246    232.100000
219    237.800000
11     238.700000
50     238.900000
441    239.100000
499    239.600000
439    242.000000
487    242